src/D2L.Bmx/OktaAuthenticator.cs

```diff
@@ -130,6 +143,129 @@ private bool TryAuthenticateFromCache(
+				var url = new Uri( page.Url );
+				if( url.Host == baseUrl.Host ) {
+					string title = await page.GetTitleAsync().WaitAsync( cancellationTokenSource.Token );
+					// DSSO can sometimes takes more than one attempt.
+					// If the path is '/' with 'sign in' in the title, it means DSSO is not available and we should stop retrying.
+					if( title.Contains( "sign in", StringComparison.OrdinalIgnoreCase ) ) {
+						if( attempt < 3 && url.AbsolutePath != "/" ) {
+							attempt++;
+							await page.GoToAsync( orgBaseAddress ).WaitAsync( cancellationTokenSource.Token );
+						} else {
+							consoleWriter.WriteWarning(
+								"WARNING: Could not authenticate with Okta using Desktop Single Sign-On." );
+							sessionIdTcs.SetResult( null );
+						}
+						return;
+					}
+				}
+				var cookies = await page.GetCookiesAsync( orgBaseAddress ).WaitAsync( cancellationTokenSource.Token );
+				if( Array.Find( cookies, c => c.Name == "sid" )?.Value is string sid ) {
+					sessionIdTcs.SetResult( sid );
+				}
+			}
+			async Task GetOktaUserEmailAsync( IResponse response ) {
+				if( response.Url.Contains( $"{orgBaseAddress}enduser/api/v1/home" ) ) {
+					string content = await response.TextAsync().WaitAsync( cancellationTokenSource.Token );
+					var home = JsonSerializer.Deserialize( content, JsonCamelCaseContext.Default.OktaHomeResponse );
+					if( home is not null ) {
+						userEmailTcs.SetResult( home.Login );
+					}
+				}
+			}
+		} catch( TaskCanceledException ) {
+			consoleWriter.WriteWarning( $"""
+				WARNING: Timed out when trying to create Okta session through Desktop Single Sign-On.
+				Check if the org '{orgBaseAddress}' is correct. If running BMX with elevated privileges,
+				rerun the command with the '--experimental-bypass-browser-security' flag
+				"""
+			);
+			return null;
+		} catch( TargetClosedException ) {
+			consoleWriter.WriteWarning( """
+				WARNING: Failed to create Okta session through Desktop Single Sign-On as BMX is likely being run
+				with elevated privileges. Rerun the command with the '--experimental-bypass-browser-security' flag.
+				"""
+			);
+			return null;
+		} catch( Exception ) {
+			consoleWriter.WriteWarning(
+				"WARNING: Unknown error while trying to authenticate with Okta using Desktop Single Sign-On." );
+			return null;
+		}
+
+		if( sessionId is null || userEmail is null ) {
+			return null;
+		} else if( !OktaUserMatchesProvided( userEmail, user ) ) {
+			consoleWriter.WriteWarning(
+				"WARNING: Could not create Okta session using Desktop Single Sign-On as "
+				+ $"provided Okta user '{StripLoginDomain( user )}' does not match user '{StripLoginDomain( userEmail )}'." );
+			return null;
+		}
+
+		var oktaAuthenticatedClient = oktaClientFactory.CreateAuthenticatedClient( orgBaseAddress, sessionId );
+		var expiresAt = ( await oktaAuthenticatedClient.GetCurrentOktaSessionAsync() ).ExpiresAt;
+		TryCacheOktaSession( user, org, sessionId, expiresAt );
```

Comment:
The only reason you passed in the `org` parameter in addition to `orgBaseAddress` is to be able to cache based on `org`.

I wonder if we'd be better off with caching by org domain/URL in general.

I don't imagine people actually doing this, but theoretically there's no harm in caching `d2l` and `d2l.okta.com` as the same thing, right?

Reply:
> The only reason you passed in the org parameter in addition to orgBaseAddress is to be able to cache based on org.

This seems reason enough to cache by orgBaseAddress